Expand README.md to cover almost everything (#713)
go-debos/debos: README.md

enough rope to do whatever tweaking which might be required behind the
scenes.

> debos expects a YAML file, called a recipe, as it's input. The recipe is

Reviewer: ```suggestion
debos expects a YAML file, called a recipe, as its input. The recipe is
```

> pre-processed by the Golang templating engine and describes a list of actions

Reviewer: ```suggestion
pre-processed by the Golang [template engine](https://pkg.go.dev/text/template) and describes a list of actions
```

> which debos runs sequentially. Actions are self-contained and independent of

Reviewer: ```suggestion
which debos runs sequentially. Actions are designed to be self-contained and independent of
```

each other.

The recipe syntax and file format and templating engine are described in the [Recipe
syntax](#recipe-syntax) section. Each available action, with its properties and a worked
example, is documented in the [Actions](#actions) section. The variables debos exports

> to the build environment and the value for the `origin` property are described in the

Reviewer: ```suggestion
to the build environment and the values for the `origin` property are described in the
```

[Environment variables](#environment-variables) section.

The reference documentation in this README is generated by hand and may lag

> behind the code. The authoritative, automatically generated action

Reviewer: ```suggestion
behind the code. The authoritative, automatically generated, action
```

documentation is published on the
[godoc documentation](https://godoc.org/github.com/go-debos/debos/actions).
When in doubt, treat the [godoc documentation](https://godoc.org/github.com/go-debos/debos/actions)

> as the more up-to-date source.

Reviewer: ```suggestion
as the most up-to-date source.
```

  it is missing (e.g. `amd64` or `arm64`).
- `sectorsize` -- logical sector size in bytes (integer). Optional, defaults
  to `512`. Set this to `4096` for devices using 4K blocks such as some UFS or
  NVMe storage. It is consumed by the `image-partition` and `raw` actions.
- `actions` -- the list of actions to run. Mandatory; at least one action is
  required.

A minimal example for a recipe:

```yaml
architecture: arm64
sectorsize: 512

actions:
  - action: debootstrap
    suite: trixie
```

### Recipe Template Variables

Two kinds of variables are passed to the recipe and they use different syntax:

- Variables passed on the command line with `-t NAME:VALUE` are fields of the
  template data object and are referenced in the recipe as `{{ .NAME }}`.
- Local variables are declared inside the recipe with
  `{{ $name := "value" }}` and referenced as `{{ $name }}`.

The common idiom combines both to give a command-line variable a default:

```yaml
{{- $image := or .image "debian.tgz" -}}

architecture: arm64

actions:
  - action: pack
    file: {{ $image }}
```

Here `debos recipe.yaml` creates `debian.tgz`, while
`debos -t image:custom.tgz recipe.yaml` creates `custom.tgz`.

> The `recipe` action passes variables down to an included recipe.

Reviewer: ```suggestion
The `recipe` action passes variables injected with `-t NAME:VALUE` down to the included recipe.
```

| ROOTDIR     | yes                    | no                   | no                  |
| IMAGEMNTDIR | yes (if image mounted) | no                   | no                  |
| IMAGE       | yes (if image exists)  | yes (if image exists)| no                  |

Example -- run a script from the recipe inside the chroot:

```yaml
- action: run
  chroot: true
  script: scripts/setup-user.sh
```

Example -- build a squashfs of the rootfs on the host:

```yaml
- action: run
  chroot: false
  command: mksquashfs ${ROOTDIR} ${ARTIFACTDIR}/root.squashfs -comp xz -e rwdata/
```

Example -- run several commands as a single script with `command: |`:

```yaml
- action: run
  chroot: true
  command: |
    apt-get update
    apt-get -y dist-upgrade
    rm -rf /var/lib/apt/lists/*
```

Example -- compress the finished image as a postprocessing step:

```yaml
{{- $image := or .image "debian.img" -}}
- action: run
  postprocess: true
  command: gzip -f {{ $image }}
```

## Other example recipes

The [doc/examples](doc/examples) directory contains a complete, runnable
example recipe [ospack-debian.yaml](/doc/examples/ospack-debian/ospack-debian.yaml) which builds a minimal

> Debian ospack tarball.

Reviewer: ```suggestion
Debian tarball.
```

> Further more detailed example recipes are stored under [debos-recipes](https://github.com/go-debos/debos-recipes) repository.

Reviewer: ```suggestion
Further more detailed example recipes are stored under the [debos-recipes](https://github.com/go-debos/debos-recipes) repository.
```

same syntax without a value. Multiple `-e` options may be given.

### Proxy configuration caveats

While the proxy variables are exported from the host to fakemachine, there
are two known sources of issues:

- `localhost` will not work from fakemachine. Use an address that is valid on
  your network. debos warns if a variable contains `localhost`, `127.0.0.1`
  or `::1`.
- Applications and scripts run inside fakemachine may expect differently
  named or differently cased proxy variables; check which variables they use.

### Variables exposed to the run action

The `run` action exposes `RECIPEDIR`, `ARTIFACTDIR`, `ROOTDIR`,
`IMAGEMNTDIR` and `IMAGE` to non-chroot commands and scripts. Which variables
are set depends on `chroot` and `postprocess`; see the table in the *run*
action above.

### Origins

Several actions accept an `origin` property that selects where their input
comes from. The following origins are always available:

- `recipe` -- the directory the recipe is in.
- `filesystem` -- the target filesystem root. Set by `filesystem-deploy` and
  `ostree-deploy` to the mounted image after deployment.
- `artifacts` -- the artifact directory.

> In addition, every `download` action registers a named origin (its `name`

Reviewer: ```suggestion
In addition, each `download` action registers a named origin (its `name`
```

debos -t image:"debian-arm64.tgz" example.yaml
```

## Recipe syntax

A recipe is a YAML file. Before it is parsed as YAML it is run through the
Golang [template engine](https://pkg.go.dev/text/template), so a recipe is
effectively a template that produces YAML. A recipe has two parts:
    - a header
    - a list of actions.

Recipe comments should start with `#`.

### Recipe Header

The recipe header may contain the following properties:

- `architecture` -- target architecture (string). Mandatory; debos aborts if
  it is missing (e.g. `amd64` or `arm64`).
- `sectorsize` -- logical sector size in bytes (integer). Optional, defaults
  to `512`. Set this to `4096` for devices using 4K blocks such as some UFS or
  NVMe storage. It is consumed by the `image-partition` and `raw` actions.
- `actions` -- the list of actions to run. Mandatory; at least one action is
  required.

A minimal example for a recipe:

```yaml
architecture: arm64

> sectorsize: 512

Reviewer: nitpick: I guess we should remove the `sectorsize` as it's optional and not needed for a `minimal example` ?

actions:
  - action: debootstrap

>     suite: trixie

Reviewer: nitpick: I guess it would make sense to add a tarball action here, to allow users to actually get some useful output if they run this recipe ?

```

### Recipe Template Variables

Two kinds of variables are passed to the recipe and they use different syntax:

- Variables passed on the command line with `-t NAME:VALUE` are fields of the
  template data object and are referenced in the recipe as `{{ .NAME }}`.
- Local variables are declared inside the recipe with
  `{{ $name := "value" }}` and referenced as `{{ $name }}`.

The common idiom combines both to give a command-line variable a default:

```yaml

> {{- $image := or .image "debian.tgz" -}}

Reviewer: ```suggestion
{{- $image := or .image "debian.tar.gz" -}}
```
nitpick - I prefer tar.gz. I know we use `tgz` elsewhere but I'd rather not introduce that in new commits ;-)

architecture: arm64

> actions:

Reviewer: nitpick: I guess it would make sense to have a `debootstrap` action here with a `$suite` variable - just to keep in line with previous example.

  - action: pack
    file: {{ $image }}
```

> Here `debos recipe.yaml` creates `debian.tgz`, while

Reviewer: see previous comment about .tar.gz

> `debos -t image:custom.tgz recipe.yaml` creates `custom.tgz`.

Reviewer: see previous comment about .tar.gz

The `recipe` action passes variables injected with `-t NAME:VALUE` down to the included recipe.
The `architecture` property of an included recipe must match the parent and

> is injected automatically as a template variable.

Reviewer: I guess an example of a subrecipe with `$architecture` set would be quite useful here ?

example recipe [ospack-debian.yaml](/doc/examples/ospack-debian/ospack-debian.yaml) which builds a minimal
Debian tarball.

Further more detailed example recipes are stored under the [debos-recipes](https://github.com/go-debos/debos-recipes) repository.

## Environment variables

### Host-propagated variables

debos reads a predefined list of environment variables from the host and
propagates them to the fakemachine build environment. The list is defined by
`environVars` in `cmd/debos/debos.go` and currently covers the proxy
variables documented at
https://wiki.archlinux.org/index.php/proxy_settings:

```
http_proxy, https_proxy, ftp_proxy, rsync_proxy, all_proxy, no_proxy
```

For each of the above variables, both the lower and upper case variant is probed
on the host and, if found, propagated to fakemachine preserving the case.
So, if the host defines `HTTP_PROXY` and `no_proxy`, both are propagated as-is.

### Setting and Overriding Variables

The `--environ-var` and `-e` options specify, overwrite or unset environment
variables for fakemachine:

```bash
debos -e ENVIRONVAR:VALUE ...
```

To unset a variable (e.g. prevent it from being propagated to fakemachine) use the

> same syntax without a value. Multiple `-e` options may be given.

Reviewer: perhaps an example may be good ?

### Proxy configuration caveats

While the proxy variables are exported from the host to fakemachine, there
are two known sources of issues:

- `localhost` will not work from fakemachine. Use an address that is valid on
  your network. debos warns if a variable contains `localhost`, `127.0.0.1`
  or `::1`.
- Applications and scripts run inside fakemachine may expect differently
  named or differently cased proxy variables; check which variables they use.

### Variables exposed to the run action

The `run` action exposes `RECIPEDIR`, `ARTIFACTDIR`, `ROOTDIR`,

> `IMAGEMNTDIR` and `IMAGE` to non-chroot commands and scripts. Which variables

Reviewer: ```suggestion
`IMAGEMNTDIR` and `IMAGE` environment variables to non-chroot commands and scripts. Which environment variables
```

I guess adding $ as a prefix to each could be good too ?
